test/test.js

```diff
@@ -0,0 +1,406 @@
+
+  const I = R.identity;
+
+  /* eslint-disable comma-dangle */
+  [
+    [2, Number, I],
+    ['2', String, I],
+    [true, Boolean, I],
+    [{a: 1}, Object, I],
+    [[0, 1, 2], Array, I],
+    [() => 1, Function, f => f()]
+  ]
+  /* eslint-enable comma-dangle */
+
+  .forEach(
+    ([expected, T, f]) => {
+      const Class = Type({T: [T]});
+      const instance = Class.T(expected);
+      const actual = instance[0];
+
+      t.deepEqual(f(expected), f(actual));
+    }
+  );
+
+  t.end();
+});
+
+test('defining a record type', t => {
+  const Point = Type(
+    {Point: {x: Number, y: Number}}
+  );
+
+  const [x, y] = Point.Point(2, 3);
+
+  const [x1, y1] = Point.PointOf({x: 2, y: 3});
+
+  t.deepEqual([x, y], [x1, y1]);
+  t.end();
+});
+
+test('create instance methods', t => {
+
+  /* eslint-disable */
+  const Maybe = Type({
+    Just: [T.Any]
+    , Nothing: []
+  });
+
+  Maybe.prototype.map = function(fn) {
+    return Maybe.case({
+      Nothing: () => Maybe.Nothing()
+      , Just: (v) => Maybe.Just(fn(v))
+    }, this);
+  }
+  /* eslint-enable */
+
+  const just = Maybe.Just(1);
+  const nothing = Maybe.Nothing();
+  just.map(R.add(1)); // => Just(2)
+
+  t.equal(nothing.map(R.add(1))._name, 'Nothing');
+  t.equal(Maybe.Just(4)[0], 4);
+  t.end();
+});
+
+
+test('create instance methods declaratively', t => {
+
+  /* eslint-disable */
+  const Maybe = Class(
+    'Maybe'
+    ,{ Just: [T.Any]
+    , Nothing: []
+    }
+    ,{
+      map( fn ){
+        return Maybe.case({
+          Nothing: () => Maybe.Nothing()
+          , Just: (v) => Maybe.Just(fn(v))
+        }, this)
+      }
+    }
+  );
+
+  /* eslint-enable */
+
+  const just = Maybe.Just(1);
+  const nothing = Maybe.Nothing();
+
+  just.map(R.add(1)); // => Just(2)
+
+  t.equal(nothing.map(R.add(1))._name, 'Nothing');
+  t.equal(Maybe.Just(4)[0], 4);
+  t.end();
+});
+
+test('Fields can be described in terms of other types', t => {
+  const Point = Type(
+    {Point: {x: Number, y: Number}}
+  );
+
+  const Shape = Type({
+    Circle: [Number, Point],
+    Rectangle: [Point, Point],
+  });
+
+
+  const [radius, [x, y]] = Shape.Circle(4, Point.Point(2, 3));
+
+  t.deepEqual([radius, x, y], [4, 2, 3]);
+
+
+  t.end();
+});
+
+test('The values of a type can also have no fields at all', t => {
+  const NotifySetting = Type(
+    {Mute: [], Vibrate: [], Sound: [T.Number]}
+  );
+
+  t.equal('Mute', NotifySetting.Mute()._name);
+
+  t.end();
+});
+
+test('If a field value does not match the spec an error is thrown', t => {
+
+  const err =
+`TypeError: Invalid value
+
+Point.Point :: Number -> Number -> { x :: Number, y :: Number }
+                         ^^^^^^
+                           1
+
+1)  "foo" :: String
+
+The value at position 1 is not a member of ‘Number’.
+`;
+
+  const Point = Named(
+    'Point'
+    , {Point: {x: Number, y: Number}}
+  );
+
+  try {
+    Point.Point(4, 'foo');
+  } catch (e) {
+    t.equal(err, e.toString());
+  }
```

Comment:
I'm not sure about this approach. Won't the assertion only be performed if evaluating the `try` block results in an exception? I think we need [`assert.throws`][1] here.

We could define a function such as this:

```javascript
//    throws :: (Function, TypeRep a, String) -> Undefined !
const throws = (f, type, message) => {
  assert.throws(f, err => err.constructor === type && err.message === message);
};
```

We'd use it like so:

```javascript
throws(() => { Point.Point(4, 'foo'); },
       TypeError,
       '...');
```


[1]: https://nodejs.org/api/assert.html#assert_assert_throws_block_error_message